django/thunderstore/api/cyberstorm/views/package_preview.py

```diff
@@ -0,0 +1,288 @@
+
+        params: OrderedDict = qp.validated_data
+        listings_qs = self.get_queryset(params)
+        listings_qs = self.order_queryset(params["ordering"], listings_qs)
+        categories = PackageCategory.objects.filter(
+            pk__in=listings_qs.aggregate(arr=ArrayAgg("categories", distinct=True))[
+                "arr"
+            ]
+        )
+        listings_page = self.paginate(params, listings_qs, params["page_size"])
+        listing_count = len(listings_qs)
+        serializer = self.serialize_results(
+            listings_page, categories, listing_count, params["page_size"]
+        )
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def get_full_cache_vary(self, params: OrderedDict) -> str:
+        cache_vary = "package_cache_key"
+        if params.get("community_identifier", None):
+            cache_vary += f".{params['community_identifier']}"
+        if params.get("namespace_identifier", None):
+            cache_vary += f".{params['namespace_identifier']}"
+        if params.get("team_identifier", None):
+            cache_vary += f".{params['team_identifier']}"
+        if params.get("user_identifier", None):
+            cache_vary += f".{params['user_identifier']}"
+        if params.get("package_identifier", None):
+            cache_vary += f".{params['package_identifier']}"
+        cache_vary += f".{params['deprecated']}"
+        cache_vary += f".{params['nsfw']}"
+        cache_vary += f".{sorted(params['included_categories'])}"
+        cache_vary += f".{sorted(params['excluded_categories'])}"
+        cache_vary += f".{params.get('section', '-')}"
+        cache_vary += f".{params.get('q', '-')}"
+        cache_vary += f".{params['ordering']}"
+        cache_vary += f".{params['page']}"
+        cache_vary += f".{params['page_size']}"
+        return cache_vary
+
+    def get_queryset(self, params) -> QuerySet[PackageListing]:
+        # Prep Q objects
+        filter_q = Q()
+        exclude_q = Q()
+
+        # Add majority of filters
+        if params.get("community_identifier", None):
+            filter_q.add(Q(community__identifier=params["community_identifier"]), Q.AND)
+
+        if params.get("namespace_identifier", None):
+            filter_q.add(
+                Q(package__namespace__name=params["namespace_identifier"]), Q.AND
+            )
+
+        if params.get("team_identifier", None):
+            filter_q.add(Q(package__owner__name=params["team_identifier"]), Q.AND)
+
+        if params.get("user_identifier", None):
+            filter_q.add(
+                Q(
+                    package__owner__members__user__username__icontains=params[
+                        "user_identifier"
+                    ]
+                ),
+                Q.AND,
+            )
+
+        if params.get("package_identifier", None):
+            filter_q.add(Q(package__name=params["package_identifier"]), Q.AND)
```

Comment:
This view will be used in different pages. Basically where ever we'll have the package listing search component. So community, user and team pages. In addition it can be used in the front page and so on.

At first I added views for each, but it kind of made no sense for every page to have it's own identical view.

Reply:
It does make sense, which comes in the form of simpler code and optimization to specific use cases. You're making quite a lot of assumptions about how those different pages work here, but I can already tell it's unlikely this implementation would meet the actual requirements.

Keeping the common filters re-usable is a good idea, but that can be achieved e.g. with a mixin class or some other utility functions. Forcing everything to go through the same view will make that one view harder to properly implement and maintain than having dedicated views for each use case.

Reply:
So should I separate this endpoint into page specific ones or keep this for now?

Reply:
Separate to page specific ones. Team and user pages are likely going to be more than just a package list query for example, and support user/team controlled pins & visibility rules etc. at some point.

Reply:
Hmm, alright. I'll try to do the separating in a clean way, split to smaller PRs, but the base code won't change that much. So please review query logic now, so that I can then just split the multiple endpoints, which all use the same base query logic from an inherited class or something.